### ISSUE TITLE
[$250] IOU - Pay button changes to Approve briefly after submitting

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#39959]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v1.4.61-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** sarahhailu1268+u008082343@gmail.com
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Issue reported by:** Applause-Internal team

Note: The original bug was fixed by another issue, but we found another [issue here](https://github.com/Expensify/App/issues/39959#issuecomment-2113295912) and will focus on fixing it.

## Action Performed:
1. Go to a workspace chat
3. Create a request money
4. Quickly click Submit
5. Verify the button changes to a pay button
6. Wait for a moment

## Expected Result:
The pay button remains displayed

## Actual Result:
It changes to an Approve button, then back to a pay button

## Workaround:
Ignore

## Platforms:

Which of our officially supported platforms is this issue occurring on?
All
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/26260477/fa43e3c1-ad09-4c42-a39b-40f046427c24

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01e2fee3e0a8646cc6</li>
        <li>Upwork Job ID: 1779912306769190912</li>
        <li>Last Price Increase: 2024-04-22</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @MitchExpensify</details>

**Comment by izarutskaya** [comment]
We think this issue might be related to the #collect project.

**Comment by sakluger** [comment]
@izarutskaya, if you're requesting money, then it doesn't make sense for the button to change to "Pay with Expensify". If you're requesting money, you would never pay. I've updated the expected behavior accordingly.

**Comment by sakluger** [comment]
Actually, I'm not sure how to get the Submit button to even appear - when I test requesting money on our own NewDot workspace, there is no option to Submit.

**Comment by BhuvaneshPatil** [comment]
can you please confirm if the person making Money Request is admin of that workspace?

And test with following scenario - Member of workspace add Money Request, can he see submit button?

**Comment by kaushiktd** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

IOU - Pay with Expensify changes back to Submit when clicking Submit button.

### What is the root cause of that problem?

When a new expense is created and the report is still being processed, displaying a pending status like `createChat,` on console log,  the submit button will persist until the report is updated.

### What changes do you think we should make in order to solve the problem?

You need to check report has `pendingAction` available or not, if not then update `shouldShowSubmitButton` condition [here](https://github.com/Expensify/App/blob/main/src/components/ReportActionItem/ReportPreview.tsx#L346) as mention below:

```
shouldShowSubmitButton && !iouReport?.pendingAction
```

### What alternative solutions did you explore? (Optional)

You could implement a popup modal that appears when the user taps on the submit button and the report is still pending action. This modal could contain a message informing the user that the report is still being processed and advising them to wait for completion. Additionally, you could provide an option for the user to dismiss the modal or continue waiting for the report to finish processing.

**Comment by sakluger** [comment]
I believe we're still waiting for proposals.

**Comment by sakluger** [comment]
I'm going to assign a new C+ since we have a few proposals waiting on review.

**Comment by c3024** [comment]
Backend resets the state and status of the report to `OPEN` so the `Submit` button appears again.

RCA and fix of disabling the button till backend returns the response suggested  by @GandalfGwaihir [here](https://github.com/Expensify/App/issues/39959#issuecomment-2057655606) looks good to me.

🎀 👀 🎀 C+ Reviewed

**Comment by kaushiktd** [comment]
@c3024 For my personal learning purpose, may I know the reason where my proposal fell behind? Curious to know if you checked my optional solution!

**Comment by sakluger** [comment]
Sent Neil a DM.

**Comment by neil-marcellini** [comment]
Thank you Sasha, it's so easy to miss these since they show as grayed out on the K2 dashboard.

**Comment by neil-marcellini** [comment]
I asked Sasha to clarify the expected result since I'm confused by [this comment](https://github.com/Expensify/App/issues/39959#issuecomment-2057316537).

**Comment by sakluger** [comment]
@GandalfGwaihir I want to make sure I understand what you're saying above. Are you saying that as the workspace owner, if I request money from someone else, I would see a "Pay with Expensify" button on my request? Why would I see the option to pay if I'm the one requesting to be paid?

**Comment by sakluger** [comment]
> Currently when the user submits expense while there is a pending action, the submit action is not getting completed and the user again has to submit the expense when there are no pending fields

That sounds like the real issue here from the customer's perspective, thanks for clarifying. @neil-marcellini does that help clarify the expected result for you?

**Comment by neil-marcellini** [comment]
Ok thanks guys. I updated the title and description according to that. Now that we agree on the problem here we can start evaluating proposals again.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by c3024** [comment]
The bug still exists

https://github.com/Expensify/App/assets/102477862/ba3fcb3c-514e-4d01-846f-566024b8b449

@GandalfGwaihir 's proposal fixes this and [my review is here](https://github.com/Expensify/App/issues/39959#issuecomment-2071396635)

**Comment by sakluger** [comment]
@neil-marcellini I updated the assignment star to be yours since we need you to re-review the proposal and assign. Thanks!

**Comment by neil-marcellini** [comment]
Thanks Sasha, it's easy to miss it otherwise. I'll review today.

**Comment by neil-marcellini** [comment]
> @GandalfGwaihir 's proposal fixes this and [my review is here](https://github.com/Expensify/App/issues/39959#issuecomment-2071396635)

Thanks for pointing out where your approval was @cubuspl42, it would be hard to find otherwise. Reviewing the [proposal](https://github.com/Expensify/App/issues/39959#issuecomment-2057655606), I don't agree with the RCA.

> Now we submit on the `FE` but when we call the backend to submit the same, we still have `createChat` as a pending field and hence the `BE` rejects the request and resets the `FE` to `submit` button.

The backend doesn't reject any request, both respond with jsonCode 200. I'm pretty sure the CreateMoneyRequest returns an open expense report, which overrides the optimistic change that made expense report submitted, and then the SubmitReport response comes back and marks it submitted again. For some reason it looks like the button doesn't update back to show as a pay button until the expense report is opened.

Anyways, instead of hiding or disabling the submit button, I think we should set some optimistic data and always show the pay button if there's a pending action to submit the report. That seems more in line with our principle of allowing the user to do as much optimistically as they can. What do you think @c3024 @GandalfGwaihir?

**Comment by cubuspl42** [comment]
I assume it's an accidental mention

**Comment by neil-marcellini** [comment]
Ok interesting, let me take another look. Last time I couldn't test fully since some other issue got in the way. If what you're saying is true, then I think we have a problem with the Network layer's SequentialQueue. It should send API write commands one at a time, so it wouldn't be possible to SubmitReport before CreateMoneyRequest is complete.

**Comment by neil-marcellini** [comment]
Ok, so I heard elsewhere that there was a problem with the SequentialQueue on staging, and it has been fixed since. While testing on dev I encountered a different, more minor problem. Unless there's already an existing issue for the new problem, I will update this issue to focus on it.

Our backend is super slow right now so I can't really test effectively on staging, but this video does prove that the SequentialQueue is working properly again; because it processes one write request at a time.


https://github.com/Expensify/App/assets/26260477/fa43e3c1-ad09-4c42-a39b-40f046427c24


https://github.com/Expensify/App/assets/26260477/05ded857-9e4c-4dee-abc1-e019b1c566d9

**Comment by sakluger** [comment]
Agreed, this can be tested more next week. I'm headed OOO tomorrow until May 31, so I've reassigned to @MitchExpensify.

If this is not finished by the time I'm back, I'll take back over then!

**Comment by MitchExpensify** [comment]
What are our next steps here @neil-marcellini ?

**Comment by neil-marcellini** [comment]
We need to retest the bug on staging to see if it matches what I found on dev as [noted here](https://github.com/Expensify/App/issues/39959#issuecomment-2113295912). If there's some other behavior on staging we should update the issue description. Then we need new proposals.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by MitchExpensify** [comment]
I cannot reproduce this on staging currently

**Comment by neil-marcellini** [comment]
If it's not reproducible then we can probably close it. Please re-open if needed.
